heat/nn/data_parallel.py

```diff
@@ -0,0 +1,275 @@
+import bisect
+import functools
+import heat as ht
+import operator
+import torch
+import torch.nn as tnn
+
+from collections import OrderedDict
+from typing import Callable, List
+
+from heat.core.communication import MPI
+
+__all__ = ["DataParallel"]
+
+
+class DataParallel(tnn.Module):
+    """
+    Implements data parallelism across multiple processes. This means that the same model will be run locally
+    on each process. Creation of the model parallels to PyTorch, the only changes are using HeAT layers (ht.nn.layer)
+    in the initialization of the network. I.E. (example code contains)
+    .. code-block:: python
+
+        class TestingModel(torch.nn.Module):
+            def __init__(self):
+                super(TestingModel, self).__init__()
+                self.net1 = ht.nn.Linear(10, 10)
+                self.relu = ht.nn.ReLU()
+                self.net2 = ht.nn.Linear(10, 5)
+
+            def forward(self, x):
+                return self.net2(self.relu(self.net1(x)))
+
+        ht_model = ht.nn.DataParallel(TestingModel(), comm, optimizer)
+
+    and a requirement of giving a HeAT communicator (``comm``) as well as a Torch optimizer (``optimizer``). For the
+    given model both the ``__init__()`` and ``forward()`` functions must be defined in the class defining the network.
+
+    It is highly recommended that a HeAT DataLoader is used, see :func:`..utils.data.datatools.DataLoader`. The
+    default communications scheme for this is blocking. The blocking scheme will average the model parameters during
+    the backwards step, synchronizing them before the next model iteration.
+
+    To use the non-blocking communications for parameter updates, negate the optional flag ``blocking``.
+
+    Attributes
+    ----------
+    module : torch.nn.Module
+        the local module
+    comm : heat.MPICommunicator
+        HeAT communicator to use
+    optimizer : torch.optim.Optimizer
+        Optimizer used for parameter updates.
+    blocking : bool (optional)
+        Flag for blocking synchronization. If not given, synchronization is blocking by default.
+    """
+
+    def __init__(
```

Comment:
im not sure how registering the hooks would work from the optimizer for this. @lehr-fa would know more (he is otherwise incapacitated today)

Reply:
We have to distinguish between two steps here: backward and optimization. Data parallel backward only accumulates gradients, which happens in the hook functions. Therefore, hooks should be part of the `DataParallel` class, as they currently are. 
I agree that your design is closer to the pytorch design. Since invoking `Optimizer.step()` and therewith performing optimization doesn't necessarily have to part of `DataParallel` class, it'll be easy to source out into a separate `ParallelOptimizer` class.

Reply:
In my proposal the dp_optimizer has access to the parallel model and can modify things there if required.

But as I said, this is a matter of taste and if things are too complicated to change or do not make sense, just let the design as it currently is.

Reply:
Actually it's not too complicated. The problem is that the relationship 'dp_optimizer knows dp_model' isn't sufficient or even necessary. We need the relationship 'dp_model knows dp_optimizer' though. The question is, how to pass dp_optimizer to dp_model in an optimal way (from design perspective)? More than ever, if there might be several dp_optimizers to be assigned to a single dp_model. I suggest to implement a simple `add_optimizers()` function, which takes either one or multiple dp_optimizers to be appended to a list attribute of dp_model.

Reply:
this is being worked on at the moment.